packages/jupyter-ai/jupyter_ai/personas/jupyternaut/jupyternaut.py

```diff
@@ -0,0 +1,56 @@
+from typing import Any
+
+from jupyterlab_chat.models import Message
+from langchain_core.output_parsers import StrOutputParser
+from langchain_core.runnables.history import RunnableWithMessageHistory
+
+from ...history import YChatHistory
+from ..base_persona import BasePersona, PersonaDefaults
+from .prompt_template import JUPYTERNAUT_PROMPT_TEMPLATE, JupyternautVariables
+
+
+class JupyternautPersona(BasePersona):
+    """
+    The Jupyternaut persona, the main persona provided by Jupyter AI.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @property
+    def defaults(self):
+        return PersonaDefaults(
+            name="Jupyternaut",
+            avatar_path="/api/ai/static/jupyternaut.svg",
+            description="The standard agent provided by JupyterLab. Currently has no tools.",
+            system_prompt="...",
+        )
+
+    async def process_message(self, message: Message):
+        provider_name = self.config.lm_provider.name
+        model_id = self.config.lm_provider_params["model_id"]
+
+        runnable = self.build_runnable()
+        variables = JupyternautVariables(
+            input=message.body,
+            model_id=model_id,
+            provider_name=provider_name,
+            persona_name=self.name,
+        )
+        variables_dict = variables.model_dump()
+        reply_stream = runnable.astream(variables_dict)
+        await self.forward_reply_stream(reply_stream)
+
+    def build_runnable(self) -> Any:
+        # TODO: support model parameters. maybe we just add it to lm_provider_params in both 2.x and 3.x
+        llm = self.config.lm_provider(**self.config.lm_provider_params)
+        runnable = JUPYTERNAUT_PROMPT_TEMPLATE | llm | StrOutputParser()
+
+        runnable = RunnableWithMessageHistory(
+            runnable=runnable,  #  type:ignore[arg-type]
+            get_session_history=lambda: YChatHistory(ychat=self.ychat, k=2),
+            input_messages_key="input",
+            history_messages_key="history",
+        )
```

Comment:
nit: would just do `return RunnableWithMessageHistory` to avoid redefining `runnable`

Reply:
I prefer always assigning return values to a local variable to allow us to log the return value before returning. Since `runnable` is already allocated on the stack here, this convenience is free.